### PR TITLE
chore/no-ref: v0.3.3 code quality, CI improvements, and test coverage

### PR DESCRIPTION
**Pull request by stephen-kintsugi**
## Summary

We haven't touched this repo in a while, so we ran a comprehensive review to address accumulated tech debt and bring CI/CD up to Kintsugi standards.

* Refactored `hook.py`: extracted `_lint_file()` from `main()`, converted `RevisionInfo` to frozen dataclass, added type hints and docstrings to all functions, added `OSError` handling for unreadable migration files
* Expanded test suite from 35 to 49 tests (97% coverage): added `test_generate_sql.py` for direct unit tests, shared `conftest.py` for fixtures, new tests for missing squawk binary, multi-file invocation, and temp path rewriting
* CI improvements aligned with kintsugi-rules patterns: conditional version bump and release (skipped for docs/test-only changes), shared composite action for Python/Poetry setup with dependency caching, SHA-pinned actions, concurrency with cancel-in-progress, `poetry version -s` and `tomllib` replacing fragile `grep | sed`, timeout-minutes on all jobs
* Pinned dev dependencies to compatible release (`~=`) from lock file versions, added `pytest-cov`
* Updated README version references to v0.3.3, added versioning docs to Local Development section
* Ran `poetry update` to pull latest compatible dependency versions

## Jira

N/A

## Tests

49 tests passing, 97% line coverage on `squawk_alembic/hook.py`. Pre-commit hooks (ruff, pyrefly, formatting) all green.

## Document Checklist

* [x] README updated with current version and versioning workflow

---

> [!NOTE]
> **Medium Risk**
> Medium risk because it refactors the core `squawk_alembic` hook execution path and adjusts CI release automation; behavior changes are mostly covered by expanded tests but could affect migration linting and tagging in edge cases.
> 
> **Overview**
> Bumps the package to **v0.3.3** and refreshes dev tooling by pinning dev dependencies (plus adding `pytest-cov`) and updating the lockfile.
> 
> Refactors `squawk_alembic/hook.py` to make linting more robust and testable: converts `RevisionInfo` to a frozen dataclass with type hints, extracts per-file lint logic into `_lint_file()`, adds clearer handling for unreadable migration files, missing `squawk`, and rewrites squawk output paths from temp files back to the original migration path.
> 
> Overhauls CI to be more deterministic and release-safe by introducing a reusable Python/Poetry setup composite action (with `.venv` caching), SHA-pinning actions, adding concurrency/timeouts, and *gating* version checks, auto-tagging, and releases on detected changes to distributable package files; version parsing is switched from `grep` to `poetry version`/`tomllib`.
> 
> Expands test coverage with shared fixtures (`tests/conftest.py`) and new unit tests for `generate_sql`, plus additional main-path tests for multi-file runs and error cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a65eaa070f6cf76e069e12d391b5d5b9af29d06f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Documentation**
  * Updated README with version references, new versioning guidelines, and release behavior.

* **Chores**
  * Bumped package to 0.3.3 and pinned dev dependency ranges.
  * Added a reusable Python/Poetry setup action and updated CI to use it; introduced concurrency, change-detection gating, and safer release/tagging steps.

* **Refactor**
  * Improved CLI hook typing and reworked linting flow for clearer skipping, error reporting, and install hints.

* **Tests**
  * Added shared fixtures and expanded tests for SQL generation and CLI scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/kintsugi-tax/squawk-pre-commit/pull/6?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .github/actions/setup-python-poetry/action.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      uses: snok/install-poetry@76e04a911780d5b312d89783f7b1cd627778900a # v1.4.1
+      with:
+        version: "2.1.3"
+        virtualenvs-create: true
+        virtualenvs-in-project: true
+        installer-parallel: true
+
+    - name: Cache virtualenv
+      if: inputs.install-dependencies == 'true'
+      uses: actions/cache@0057852bfaa89a56745cba8c7296529d2fc39830 # v4.3.0
```
Comment (stephen-kintsugi) [inline .github/actions/setup-python-poetry/action.yml:32]: **Shared composite action for Python/Poetry setup**

Extracted from the inline steps that were duplicated between `lint-and-test` and `auto-tag` jobs. Follows the kintsugi-rules pattern: SHA-pinned actions, `poetry sync -n`, `installer-parallel: true`, and a cache key that includes the Python version for correctness across version bumps. The `install-dependencies` input lets the `auto-tag` job skip `poetry install` since it only needs `poetry version -s`.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -6,40 +6,57 @@ on:
...
-          python-version: '3.12'
+          fetch-depth: 0
 
-      - name: Install Poetry
-        uses: snok/install-poetry@76e04a911780d5b312d89783f7b1cd627778900a # v1.4.1
-        with:
-          version: "2.1.3"
-          virtualenvs-create: true
-          virtualenvs-in-project: true
+      - name: Detect code changes
```
Comment (stephen-kintsugi) [inline .github/workflows/ci.yml:26]: **Conditional version bump and release based on package file changes**

Previously, every PR (including docs and test changes) required a version bump and triggered a release on merge. Now we diff against `squawk_alembic/`, `pyproject.toml`, and `.pre-commit-hooks.yaml` only. The `code_changed` output gates both the version consistency check (line 56) and the `auto-tag` job (line 84). Lint and tests still run on every change.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -6,40 +6,57 @@ on:
...
       - name: Check version consistency
+        if: steps.changes.outputs.code_changed == 'true'
         run: |
-          PYPROJECT_VERSION=$(grep '^version = ' pyproject.toml | sed 's/version = "\(.*\)"/\1/' || true)
-          INIT_VERSION=$(grep '^__version__ = ' squawk_alembic/__init__.py | sed 's/__version__ = "\(.*\)"/\1/' || true)
-          if [ -z "$PYPROJECT_VERSION" ] || [ -z "$INIT_VERSION" ]; then
-            echo "::error::Could not parse version from pyproject.toml or __init__.py"
-            exit 1
-          fi
+          PYPROJECT_VERSION=$(poetry version -s)
```
Comment (stephen-kintsugi) [inline .github/workflows/ci.yml:58]: **Replaced fragile `grep | sed` with `poetry version -s` and `tomllib`**

The old version parsing (`grep '^version = ' pyproject.toml | sed ...`) would break if another TOML table had a `version` key. `poetry version -s` is authoritative for the current version, and `tomllib` (stdlib since 3.11) properly parses the TOML from `git show` for the main branch comparison.

File: squawk_alembic/hook.py [hunk 4]
```diff
@@ -36,50 +37,50 @@ def find_migrations_path():
     return None
 
 
+@dataclass(frozen=True, slots=True)
 class RevisionInfo:
```
Comment (stephen-kintsugi) [inline squawk_alembic/hook.py:41]: **`RevisionInfo` converted to frozen dataclass**

Was a manual class with `__slots__` and hand-written `__init__`. The frozen dataclass gives us `__repr__` and `__eq__` for free (helps with debugging and test assertions), enforces immutability, and the type annotations on the fields document the `down_revision` union (`str | tuple[str, ...] | None`) which was previously implicit.

File: squawk_alembic/hook.py [hunk 5]
```diff
@@ -187,7 +188,66 @@ def file_exists_on_branch(filepath, branch):
     return result.returncode == 0
 
 
-def main():
+class _SquawkNotFound(Exception):
+    pass
+
+
+def _lint_file(filepath: str, migrations_path: Path, diff_branch: str | None) -> int:
```
Comment (stephen-kintsugi) [inline squawk_alembic/hook.py:195]: **Extracted per-file logic from `main()` into `_lint_file()`**

`main()` was 73 lines mixing argument parsing, file iteration, SQL generation, temp file management, and subprocess calls. This extraction isolates the generate-write-lint-cleanup cycle for a single file, making `main()` a clean orchestration loop. The `_SquawkNotFound` exception bubbles up the "squawk binary missing" case, which should halt all processing (not just one file).

**Review by stephen-kintsugi — COMMENTED** [review]
Self-review walkthrough of key changes in this PR.

File: squawk_alembic/hook.py [hunk 6]
```diff
@@ -187,7 +188,66 @@ def file_exists_on_branch(filepath, branch):
...
+
+    if diff_branch and file_exists_on_branch(filepath, diff_branch):
+        return 0
+
+    try:
+        sql = generate_sql(filepath)
+    except GenerateSqlError as exc:
+        print(str(exc), file=sys.stderr)
+        return 1
+    except OSError as exc:
```
Comment (stephen-kintsugi) [inline squawk_alembic/hook.py:213]: **`OSError` handling for unreadable migration files**

Previously, a permission error on a migration file would crash with a raw traceback. Now it produces a consistent `squawk-alembic:` prefixed message. The error string from `OSError` already includes the filepath, so we avoid duplicating it in the prefix.

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    """Set up a fake repo with alembic config and a versions directory."""
+    monkeypatch.chdir(tmp_path)
+    versions = tmp_path / "migrations" / "versions"
+    versions.mkdir(parents=True)
+    (tmp_path / "alembic.ini").write_text("[alembic]\nscript_location = ./migrations\n")
+    return tmp_path
+
+
+def make_result(returncode=0, stdout="", stderr=""):
+    return SimpleNamespace(returncode=returncode, stdout=stdout, stderr=stderr)
```
Comment (stephen-kintsugi) [inline tests/conftest.py:20]: **Shared test fixtures with `SimpleNamespace`**

Consolidated duplicate `repo` fixtures, `write_migration`, `fake_subprocess`, and `make_result` from `test_main.py`. `make_result` was using `type('Result', (), {...})()` to create anonymous classes on every call; `SimpleNamespace` is the idiomatic stdlib choice for this.

File: tests/test_main.py [hunk 8]
```diff
@@ -134,6 +82,46 @@ def upgrade():
     assert "some squawk warning" in captured.out
 
 
+def test_squawk_failure_replaces_tmp_path_in_output(repo, capsys):
```
Comment (stephen-kintsugi) [inline tests/test_main.py]: **New test: temp file path rewriting in squawk output**

Verifies that `hook.py:234-235` replaces the temp file path with the original migration path in both stdout and stderr. The mock squawk process returns output containing the temp path, and we assert the user sees the original filepath instead.

File: tests/test_main.py [hunk 9]
```diff
@@ -236,6 +279,95 @@ def upgrade():
...
+    with (
+        patch("sys.argv", ["squawk-alembic", path1, path2]),
+        patch("subprocess.run", side_effect=fake_subprocess()) as mock_run,
+    ):
+        assert main() == 0
+        # alembic + squawk for each file = 4 calls
+        assert mock_run.call_count == 4
+
+
+def test_multiple_files_first_fails_second_still_runs(repo, capsys):
```
Comment (stephen-kintsugi) [inline tests/test_main.py]: **New test: multi-file invocation with partial failure**

All previous tests passed a single file. This verifies the exit-code accumulation logic: when the first file fails (alembic error), the hook continues processing the second file and still returns exit code 1. Previously this code path had zero test coverage.

File: pyproject.toml [hunk 10]
```diff
@@ -29,11 +29,12 @@ python = ">=3.10"
 squawk-cli = ">=2.0"
 
 [tool.poetry.group.dev.dependencies]
-alembic = "*"
-pre-commit = "*"
-pytest = "*"
-ruff = "*"
-pyrefly = "*"
+alembic = "~=1.18.4"
```
Comment (stephen-kintsugi) [inline pyproject.toml:32]: **Dev dependencies pinned to compatible release**

Were all `"*"` (unpinned). Now use `~=` pinned to the bugfix level from `poetry.lock`, so `poetry update` allows patches but not breaking minor/major bumps. Runtime dep `squawk-cli` stays at `>=2.0` intentionally since consumers control their squawk version via `additional_dependencies` in their pre-commit config.

**Review by michael-kintsugi — DISMISSED** [review]

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -6,40 +6,57 @@ on:
...
     steps:
       - name: Checkout code
-        uses: actions/checkout@v4
-
-      - name: Set up Python
-        uses: actions/setup-python@v5
+        uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
         with:
-          python-version: '3.12'
+          fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (stephen-kintsugi) [reply .github/workflows/ci.yml:25]: We don't set persist-credentials: false anywhere. The default (true) is correct here since lint-and-test runs git fetch and auto-tag runs git push.

File: tests/test_main.py [hunk 12]
```diff
@@ -162,6 +150,29 @@ def upgrade():
...
+        def upgrade():
+            pass
+        """,
+    )
+    import os
+
+    os.chmod(repo / "migrations" / "versions" / "026_unreadable.py", 0o000)
+    with patch("sys.argv", ["squawk-alembic", path]):
+        assert main() == 1
+    os.chmod(repo / "migrations" / "versions" / "026_unreadable.py", 0o644)
```
[the remark this answers is not in the document]
Comment (stephen-kintsugi) [reply tests/test_main.py]: Fixed in 9dc608f. Wrapped in try/finally so permissions are restored even if the assertion fails.

**Review by michael-kintsugi — APPROVED** [review]
